ImperatorToCK3Tests/MapperTests/RegionMapper/CK3RegionMapperTests.cpp

```diff
@@ -0,0 +1,232 @@
+	theMapper.loadRegions(landedTitles, regionStream, islandRegionStream);
+
+	ASSERT_EQ("c_athens", *theMapper.getParentCountyName(79));
+	ASSERT_EQ("d_wakaba", *theMapper.getParentDuchyName(79));
+	ASSERT_EQ("test_region", *theMapper.getParentRegionName(79));
+	ASSERT_EQ("c_defff", *theMapper.getParentCountyName(6));
+	ASSERT_EQ("d_hujhu", *theMapper.getParentDuchyName(6));
+	ASSERT_EQ("test_region2", *theMapper.getParentRegionName(6));
+}
+
+TEST(Mappers_RegionMapperTests, wrongParentLocationsReturnEmpty)
+{
+	mappers::CK3RegionMapper theMapper;
+	CK3::LandedTitles landedTitles;
+	std::stringstream landedTitlesStream;
+	landedTitlesStream << "d_testduchy = { 1 2 3 } \n";
+	landedTitlesStream << "d_testduchy2 = { 4 5 6 } ";
+	landedTitles.loadTitles(landedTitlesStream);
+	std::stringstream regionStream;
+	regionStream << "test_region = { duchies = { d_testduchy } }";
+	regionStream << "test_region2 = { duchies = { d_testduchy2 } }\n";
+	std::stringstream islandRegionStream;
+	theMapper.loadRegions(landedTitles, regionStream, islandRegionStream);
+
+	ASSERT_FALSE(theMapper.getParentCountyName(7));
+	ASSERT_FALSE(theMapper.getParentDuchyName(7));
+	ASSERT_FALSE(theMapper.getParentRegionName(7));
+}
+
+TEST(Mappers_RegionMapperTests, locationNameValidationWorks)
+{
+	mappers::CK3RegionMapper theMapper;
+	
+	CK3::LandedTitles landedTitles;
+	std::stringstream landedTitlesStream;
+	landedTitlesStream << "k_ugada = { d_wakaba = { c_athens = { b_athens = { province = 79 } b_newbarony = { province = 56 } } } } \n";
+	landedTitlesStream << "k_ghef = { d_hujhu = { c_defff = { b_cringe = { province = 6 } b_newbarony2 = { province = 4 } } } } \n";
+	landedTitlesStream << "c_county = { b_barony = { province = 69 } } \n";
+	landedTitles.loadTitles(landedTitlesStream);
+	
+	std::stringstream regionStream;
+	regionStream << "test_region = { duchies = { d_wakaba } }";
+	regionStream << "test_region2 = { duchies = { d_hujhu } }\n";
+	regionStream << "test_region3 = { regions = { test_region test_region2 } counties = { c_county } }\n";
+	std::stringstream islandRegionStream;
+	theMapper.loadRegions(landedTitles, regionStream, islandRegionStream);
+
+	ASSERT_TRUE(theMapper.regionNameIsValid("d_wakaba"));
+	ASSERT_TRUE(theMapper.regionNameIsValid("test_region2"));
+	ASSERT_TRUE(theMapper.regionNameIsValid("test_region3"));
+	ASSERT_TRUE(theMapper.regionNameIsValid("c_county"));
+	ASSERT_FALSE(theMapper.regionNameIsValid("nonsense"));
+}
+
+
+TEST(Mappers_RegionMapperTests, locationServicesSucceedsForProvinceField)
+{
+	mappers::CK3RegionMapper theMapper;
+	CK3::LandedTitles landedTitles;
+	std::stringstream regionStream;
+	regionStream << "test_region = { provinces = { 1 2 69 7 } }";
+	std::stringstream islandRegionStream;
+	theMapper.loadRegions(landedTitles, regionStream, islandRegionStream);
+
+	ASSERT_TRUE(theMapper.provinceIsInRegion(69, "test_region"));
+}
+
+TEST(Mappers_RegionMapperTests, locationServicesSucceedsForCountyField)
```

Comment:
you already tested this and one above with locationServicesWork. Might as well make a separate test for region and superregion fields.

Reply:
This and `locationServicesWork` are different:
- in `locationServicesWork` there is a check whether a province is in a county that's part of a duchy contained by a region
- in `locationServicesSucceedsForCountyField` we check whether a province is in a county that's **directly** contained by a region (`test_region = { counties = { c_athens } }`)
For coverage we need both.
I think there's no need for `locationServicesSucceedsForRegionsField`, in regions case this is already tested by the fourth and fifth asserts in `locationServicesWork`.

I hope this isn't too confusing 😅.

Reply:
Perfectly understandable.